core/client/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -528,20 +530,26 @@ private void updateCacheStream(long blockId) throws IOException {
       return;
     }
 
-    // If this block is read from a remote worker but we don't have a local worker, don't cache
-    if (isReadingFromRemoteBlockWorker() && !mContext.hasLocalWorker()) {
-      return;
-    }
-
     // Unlike updateBlockInStream below, we never start a block cache stream if mPos is in the
     // middle of a block.
     if (mPos % mBlockSize != 0) {
       return;
     }
 
     try {
-      WorkerNetAddress address = mLocationPolicy.getWorkerForNextBlock(
-          mBlockStore.getWorkerInfoList(), getBlockSizeAllocation(mPos));
+      // If this block is read from a remote worker, we should never cache except to a local worker.
+      if (isReadingFromLocalBlockWorker()) {
+        WorkerNetAddress localWorker = mContext.getLocalWorker();
+        if (localWorker != null) {
+          mCurrentCacheStream =
+              mBlockStore.getOutStream(blockId, getBlockSize(mPos), localWorker, mOutStreamOptions);
```

Comment:
1. I think it's simpler to only use the write policy if the block is not already cached. Then the contract for location policy can be "If the block is not yet cached in alluxio, it will be cached to the location returned by the location policy". Otherwise we have to add an addendum that we will ignore any location suggestions that would cache an already-cached block remotely.

2. I thought you were saying earlier that it's a waste of time to get a list of all workers when the only one you would consider caching to is the local one.